torch/csrc/utils/pybind.h

```diff
@@ -107,3 +107,52 @@ template<> struct type_caster<at::IntArrayRef> {
 template <typename T>
 struct type_caster<c10::optional<T>> : optional_caster<c10::optional<T>> {};
 }} // namespace pybind11::detail
+
+namespace torch {
+namespace impl {
+
+// This function is a valid shared_ptr destructor and can be used to
+// conveniently allocate a shared_ptr to an object whose destructor will be run
+// without the GIL.  Often, this logic cannot be conveniently placed on the
+// destructor itself because the object has no dependency on Python.
```

Comment:
nit: it wasn't clear to me at the beginning that this logic is for objects that are used either in a pure c++ context or a python-bound context. And so this custom destructor allows to bind the GIL-less destruction to specific instances of the object (namely the python-bound ones).

But I'm not sure how to express that idea in a small sentence :D

Reply:
I didn't have time to write a short letter so I wrote a long letter